### ISSUE TITLE
Synchronization  with tan

### ISSUE THREAD

**Issue by Zeljko97** [robocode13/lib-fints#1]
When tan is required, after first synchronize(), we get tan metod, and again synchronize() call, and after that we need to add tan code, what is need to be next, to get account or account statemens. I tried with synchronizeWithTan, or getAccountsStatmentsWithTan but i received message that 'Dialog must be initialized'. 
And if I try first with as in code below client.getAccountStatements(account) in bankingInformation I don't have upd.
What it's wrong, please help?

const client = new FinTSClient(config);
    let syncResponse = await client.synchronize(); //first sync call
    console.log('Banking information: ' + syncResponse.bankingInformation);
    console.log('Bank answers: ' + JSON.stringify(syncResponse.bankAnswers))

    if (syncResponse.success === true) {
        client.selectTanMethod(921)
        const sync = await client.synchronize() //second sync call

        if (sync.requiresTan === true) {
            const account = sync.bankingInformation.upd.bankAccounts[0];
            const response = await client.getAccountStatements(account)
            if (response.requiresTan) {
                rl.question('PIN: ', async (answer) => {
                        const response = await client.getAccountStatementsWithTan(sync.tanReference, answer)
                        console.log('Response: ' + response)
                        console.log('JSON stringify: ' + JSON.stringify(response))

                })
            }

        }
    } else {
        console.log(false)
    }

**Comment by robocode13** [comment]
Hi
when the second synchronize() call returns response.requiresTan, you have to continue this sychronize call first with a TAN by calling synchronizeWithTan(sync.tanReference, tan), only after this call will the response contain the complete bankingInformation with UPD. Then you can continue like in your example with getAccountStatements().

so roughly like this:

    ...
    let sync = await client.synchronize() //second sync call

    if (sync.requiresTan === true) {
      const tan = await rl.question(sync.tanChallenge + ': ');
      sync = await client.synchronizeWithTan(sync.tanReference, tan)

      const account = sync.bankingInformation.upd.bankAccounts[0];
      const response = await client.getAccountStatements(account)
      ...
    }


Let me know if this helps.

**Comment by Zeljko97** [comment]
Hi. I tried with client.synchronizationWithTan(sync.tanReference, tan), but again i received error message: 

lib-fints/dist/dialog.js:149
            throw Error('dialog must be initialized before sending a TAN message');
                  ^

Error: dialog must be initialized before sending a TAN message

**Comment by robocode13** [comment]
Could it be that your second sync request is not successfull, because you don't check this in your code above.
Can you try to check sync.success and sync.bankAnswers before checking sync.requiresTan

**Comment by Zeljko97** [comment]
Yes, I checked as you said, and it's success.
This is the response: 

success: true
Bank answers: [{"code":3060,"text":"Bitte beachten Sie die enthaltenen Warnungen/Hinweise."},{"code":3920,"text":"Zugelassene Zwei-Schritt-Verfahren für den Benutzer.","params":["921","922"]},{"code":30,"text":"Auftrag empfangen - Bitte die empfangene TAN eingeben.(MBT62820200002)"}]

**Comment by robocode13** [comment]
Is this the response to your second sync call or the first one?
could you please paste again your complete code as it is now.

**Comment by Zeljko97** [comment]
This is response from second call.
Yes, of course, this is my code: 

 const config = FinTSConfig.forFirstTimeUse('xxx',
        '1.0',               // Version
        'xxx',
        'xxx',           // Bank code (BLZ)
        'xxx',
        'xxx'
    );

    const client = new FinTSClient(config);
    let syncResponse = await client.synchronize();

    if (syncResponse.success === true) {
        client.selectTanMethod(921)

        let sync = await client.synchronize()

        console.log('Success: ' + sync.success)
        console.log('Bank answers: ' + JSON.stringify(sync.bankAnswers) )
        if (sync.success) {
            if (sync.requiresTan === true) {
                rl.question(sync.tanChallenge, async (answer) => {
                    sync = await client.synchronizeWithTan(sync.tanReference, answer)
                    const account = sync.bankingInformation.upd.bankAccounts[0];
                    const response = await client.getAccountStatements(account)
                })
            }
        }

    } else {
        console.log(false)
    }

**Comment by robocode13** [comment]
I think found a bug where the initialized state was not set correctly for the dialog when a TAN is required.
I just published a new package version 1.0.2. Can you update and try this new version please.

**Comment by Zeljko97** [comment]
Yes, it works. Thank you very much. 
One more question: Is it necessary to every time send tan challenge, or after this one first we can get transaction data without chalenge (pincode)?

**Comment by robocode13** [comment]
[](url)Nice to hear that it works now.
Regarding the TAN challenge, this is the decision of the bank and cannot be influenced by the client. But you shouldn't have to do the synchronization calls again after you got the UPD (account) information.
Take a look at the README file to see how: [Starting a Session from saved banking information](https://github.com/robocode13/lib-fints#starting-a-session-from-saved-banking-information)

**Comment by robocode13** [comment]
@Zeljko97 Out of curiosity, can you tell me with which bank you tested the library so i might add it to the "Successfully tested with" section. Thanks!

**Comment by Zeljko97** [comment]
@robocode13 Yes, it is 'Sparkasse Bank'. 
I have one more question. 
When I chose the first tan method, as stated in the documentation the sync went through. But I was told by the bank that that method will soon stop being used, it is the pushTan(921) method. From the bank's response, I can see that method 922 is available and when I select client.selectTanMethod(922), I get an exception: Error: given tanMethodId is not available according to the BPD. Can you please check why this is happening?

**Comment by Zeljko97** [comment]
This is whole answer from bank support: 
The log data shows that the software still uses the older segment version 6 of the HKTAN. This means that the older pushTAN 1.0 security procedure (security function=921) is used with the output of a six-digit TAN in the S-pushTAN app. This procedure will no longer be supported in the foreseeable future (date still open). You should check that the software you are using also supports segment version seven of the HKTAN. The HKTAN#7 is mandatory for the current backup procedures pushTAN 2.0 (decoupled backup function=922) and pushTAN-ZAS (backup function=923). Details can be found in the FinTS specification in the document "Security PIN/TAN".

**Comment by Zeljko97** [comment]
Hi,

Can you please check the question that I posted on github.

Thank you in advance.

чет, 31. окт 2024. у 15:41 Robert Weber ***@***.***> је
написао/ла:

**Comment by robocode13** [comment]
Yes, these "decoupled methods" are not yet supported by the library. I need to implement Version 7 of the HKTAN element. This is on my TODO list and i will try to implement it as soon as possible, but as this is a free-time project i cannot tell you exactly when this will be done.

**Comment by Zeljko97** [comment]
I understand and thank you for your quick reply.
It would be good if it also supports version 7, because the library is well written and one of the few that works with FinTs.

**Comment by Zeljko97** [comment]
@robocode13 I am trying to use now FinTsConfig.fromBankingInformation:

    const client = new FinTSClient(config);

    let sync = await client.synchronize();

    if (sync.success === true) {

            const account = sync.bankingInformation.upd.bankAccounts[0];
            const response = await client.getAccountStatements(account.accountNumber)

            console.log('Response: ' + response)
    }

and I received this message: "bankAnswers":[{"code":9050,"text":"Die Nachricht enthält Fehler."},{"code":9075,"text":"Banking-Programm nicht PSD2-fähig, bitte aktualisieren."}],"requiresTan":false}. Is it error familiar to you?

**Comment by robocode13** [comment]
when you use `fromBankingInformation` with a `bankingInformation` object you saved after previous synchronizations, you shouldn't need to call `synchronize` again, you can directly start with `getAccountStatements()`.
I am not exactly sure what the error means though, did you provide the selected TAN method either in the `fromBankingInformation()` method or by calling `client.selectTanMethod()` ?

**Comment by Zeljko97** [comment]
Is this what the code should look like now?

 const config = FinTSConfig.fromBankingInformation('xxx',
        'xx',               // Version
        bankingInformation,
        'xxx',
        'xxx'
    );

    const client = new FinTSClient(config);
    const response = await client.getAccountStatements('xxx')
    console.log('Response: ' + response)

}

**Comment by robocode13** [comment]
You always have to specify the TAN method for valid requests, the bank will decide if it requires a TAN for a transaction or not. You cannot prevent that by ommiting the TAN method selection.

**Comment by robocode13** [comment]
> Is this what the code should look like now?
> 
> const config = FinTSConfig.fromBankingInformation('xxx', 'xx', // Version bankingInformation, 'xxx', 'xxx' );
> 
> ```
> const client = new FinTSClient(config);
> const response = await client.getAccountStatements('xxx')
> console.log('Response: ' + response)
> ```
> 
> }

yes, but you also need to give the tanMethodId in the fromBankingInformation() call

**Comment by robocode13** [comment]
> I understand and thank you for your quick reply. It would be good if it also supports version 7, because the library is well written and one of the few that works with FinTs.

@Zeljko97 
I just created a new package version 1.1.0 which now supports the decoupled (v7) TAN methods. Maybe you can give it a try and give some feedback if it works.
